File: python/day8.py
def say_hello():
    print("Hello from a function")

def add(x,y):
    sum = x+y
    return sum

# Python's way of handling errors is "exceptions".
# The "try...except" block is the standard way to handle
# code that might fail.
def divide(numerator, denominator):
    if denominator == 0:
        raise ValueError("Cannot devide by zero")
    
    return numerator / denominator

say_hello()

print(f"2 + 3 is :",add(2,3))

# We must "try" to run the code that might fail.
try:
    quotient = divide(10,2)
    print(f"10 / 2 = {quotient}")

    quotient2 = divide(9,0)
    print(f"9 / 0 = {quotient2}")
except ValueError as e:
    print(f"An error occurred: {e}")

# Python *can* return multiple values (as a tuple)
def get_name():
    return "John", "Deo"

first, last = get_name()
print(f"Name is :{first} {last}")

File: go-basic/day8.go
package main

import "fmt"

// Single function with no parameters and return value
func sayHello() {
	fmt.Println("Hello from a function!")
}

// 2. A function with parameters
// Will take two parameters x and y
// we must specify their types
// and return single value

func add(x int, y int) int {
	sum := x + y
	return sum
}

// Go function can have multi return values
// This function returns two values: a 'string' and an 'error'.
// This is the standard way Go handles functions that might fail.
func devide(numerator int, denominator int) (int, bool) {
	if denominator == 0 {
		// We can't divide by zero.
		// Return 0 (as the int) and 'true' (signaling an error occurred).
		return 0, true
	}

	return numerator / denominator, false
}

// main function
func main() {
	sayHello()

	// result := add(5, 3)
	fmt.Println("5 + 3 is :", add(5, 3))

	// Call 'divide' and catch *both* return values.
	// We use the ':= ' syntax to declare two new variables.

	quotient, didError := devide(10, 2)
	if didError == true {
		fmt.Println("Error : Cannot divide by zero")
	} else {
		fmt.Println("10 / 2 =", quotient)
	}

	// You can use '_' (the blank identifier) to
	// ignore a return value you don't care about.
	_, err := devide(9, 0)
	if err {
		fmt.Println("second division failed")
	}
}
